src/librustc/ty/context.rs

```diff
@@ -1510,6 +1510,20 @@ impl<'tcx> TyCtxt<'tcx> {
                 .subst(*self, self.mk_substs([self.lifetimes.re_static.into()].iter())),
         )
     }
+
+    /// Returns a displayable description and article for the given `def_id` (e.g. `("a", "closure")`).
+    pub fn article_and_description(
+        &self,
+        def_id: crate::hir::def_id::DefId,
+    ) -> (&'static str, &'static str) {
+        self.def_kind(def_id).map_or_else(
+            || {
+                // TODO: is it a problem to try to use the ty here?
```

Comment:
Looking at what we have available in rustc I would probably start by matching on `tcx.def_key(def_id).disambiguated_data.data` and then using `def_kind` for the `TypeNs`, `ValueNs` and `MacroNs` cases.

Which leaves closures. I would rather not use the type here, since it loses information on async generators. I guess there are a few options here:

* Define a `generator_kind(DefId) -> Option<GeneratorKind>` query.
* Make a method of `MirBorrowckCtxt` that uses `generator_kind` in the MIR body. Make this query only get `closure/generator` correct.

Reply:
> Define a generator_kind(DefId) -> Option<GeneratorKind> query.

@matthewjasper How would I go about doing this?

Reply:
Change the variant of `EntryKind` to be `Generator(hir::GeneratorKind)` (the current contents are unused).
Then copy and modify `static_mutability`.

Reply:
Do you mean this one: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/rmeta/enum.EntryKind.html#variant.Generator

Reply:
@matthewjasper Could you give a bit more detail? Is there a query to get the `EntryKind` for a def_id?

Reply:
Yes, that's the correct variant.

Queries that use metadata work as follows:
1. Define a version of the query for items in the current crate https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_typeck/collect.rs#L2597-L2607
1. Define a method on `CrateMetadata` (`self.kind` will get the entry): https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_metadata/rmeta/decoder.rs#L1389-L1395
1. Add a provider for external crates that uses that method: https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_metadata/rmeta/decoder/cstore_impl.rs#L136